src/core/qgspallabeling.cpp

```diff
@@ -1199,7 +1252,7 @@ void QgsPalLayerSettings::registerFeature( const QgsFeature &f, QgsRenderContext
   {
     if ( isObstacle )
     {
-      registerObstacleFeature( f, context, labelFeature, obstacleGeometry );
+      registerObstacleFeature( feature, context, labelFeature, obstacleGeometry );
```

Comment:
Things are complex here. We need to ensure that obstacle geometry is ALWAYS to original feature geometry and never the newly generated geometry (both for this early-exit "obstacle only" path and the full labeling path). The rationale is described here:

https://github.com/qgis/QGIS-Enhancement-Proposals/issues/46#issuecomment-155097178

Basically, the obstacle geometry must ALWAYS be the original feature geometry as was used for rendering the symbol itself. Picture a bendy river which has a label geometry generator using  simplify / smooth functions to get nicer curved labels -- it's the original rendered river path which is the obstacle for other labels, not the smoothed label path (otherwise labels from other labels will overlap the parts of the rendered river which are smoothed out in the label path).

When fixing this keep in mind that the obstacleGeometry argument for registerFeature is an override, and may be a null geometry, which means "use the original feature geometry as an obstacle for other layer's labels".